server/rpc/admin_server.go

```diff
@@ -363,8 +363,16 @@ func (s *adminServer) RemoveDocumentByAdmin(
 		}
 	}()
 
-	// TODO(emplam27): Add an option to remove the document if there are no clients attached to it.
-	if err := documents.RemoveDocument(ctx, s.backend, project, docInfo.ID); err != nil {
+	isAttached, err := documents.IsAttachedDocument(ctx, s.backend, project, docInfo.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	if isAttached && !req.ForceRemoveIfAttached {
+		return nil, fmt.Errorf("remove document: document is attached")
+	}
+
+	if err := documents.RemoveDocument(ctx, s.backend, docInfo.ID); err != nil {
```

Comment:
Other clients can attach the given document between 366 and 375 lines are executing.

RemoveDocumentByAdmin acquires a lock of PushPullKey of the given document to ensure consistency, but AttachDocument only acquires when there are changes in the given request pack. 🤔

We need to think about how to handle this problem.

Reply:
As we discussed, it would be good to use locks regardless of whether or not changes exists on the request pack only on attach and detach. Since the APIs are related to start editing or finish editing and pure editing is separated by PushPull, it doesn't seem to cause a problem with concurrency.